File: fliply-back/src/main/java/com/fliply/model/CardProgress.java
package com.fliply.model;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "card_progress", uniqueConstraints = {
    @UniqueConstraint(columnNames = {"user_id", "card_id"})
})
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class CardProgress {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @com.fasterxml.jackson.annotation.JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @com.fasterxml.jackson.annotation.JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "card_id", nullable = false)
    private Card card;

    @Column(name = "right_count", nullable = false)
    @Builder.Default
    private Integer rightCount = 0;

    @Column(name = "wrong_count", nullable = false)
    @Builder.Default
    private Integer wrongCount = 0;

    @Column(name = "unsure_count", nullable = false)
    @Builder.Default
    private Integer unsureCount = 0;

    @Column(nullable = false)
    @Builder.Default
    private Integer streak = 0;

    @Column(name = "difficulty_score", nullable = false, precision = 5, scale = 2)
    @Builder.Default
    private BigDecimal difficultyScore = BigDecimal.ONE;

    @Column(name = "review_after")
    private LocalDateTime reviewAfter;

    @Column(name = "last_review_at")
    private LocalDateTime lastReviewAt;

    @Column(name = "total_reviews", nullable = false)
    @Builder.Default
    private Integer totalReviews = 0;

    @Column(nullable = false, length = 20)
    @Builder.Default
    private String status = "new";

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at", nullable = false)
    @Builder.Default
    private LocalDateTime updatedAt = LocalDateTime.now();

    @PreUpdate
    public void preUpdate() { this.updatedAt = LocalDateTime.now(); }
}

File: flashmind-front/models/CardProgress.js
import { Api } from '../utils/api.js'

export class CardProgress {
  constructor(data) {
    this.id               = data.id
    this.user_id          = data.userId
    this.card_id          = data.cardId
    this.right_count      = data.rightCount ?? 0
    this.wrong_count      = data.wrongCount ?? 0
    this.unsure_count     = data.unsureCount ?? 0
    this.streak           = data.streak ?? 0
    this.difficulty_score = data.difficultyScore ?? 1.0
    this.review_after     = data.reviewAfter
    this.last_review_at   = data.lastReviewAt
    this.total_reviews    = data.totalReviews ?? 0
    this.status           = data.status ?? 'new'
  }

  static async fetchByDeck(deckId) {
    const data = await Api.get(`/decks/${deckId}/progress`)
    return data.map(d => new CardProgress(d))
  }

  static async recordAnswer(sessionId, cardId, answerType) {
    return Api.post(`/sessions/${sessionId}/answers`, { cardId, answerType })
  }
}
